### ISSUE TITLE
Order of equally scored alignments in pairwise2.globalms (deprecated) and Align (new)

### ISSUE THREAD

**Issue by gph82** [biopython/biopython#4360]
This is not a bug, rather a question about migration pairwise2.globalms -> Align

### Setup
Biopython version, Python version, and operating system as follows:

```3.9.12 (main, Apr  5 2022, 06:56:58) 
[GCC 7.5.0]
CPython
Linux-5.15.0-76-generic-x86_64-with-glibc2.35
1.79
```
In migrating away from `pairwise2` (following deprecation notice) to `Align`, I notice the inversion of the order of equivalent alignments. 

Please note, this is a bogus edge case I've constructed in my testing suite (for reasons not relevant here) and since my test was expecting the _old_ alignment, so I found this quirk.

```
from Bio import pairwise2, Align
seq1 = "DQ"
seq2 = "TEDQRNEEKAQREANKKIDQKQLQKDKQVYR"
print("Old behavior")
[print(pairwise2.format_alignment(*a)) for a in pairwise2.align.globalms(seq1,seq2,1,0,-1,0)]
print("New behavior")
new = Align.PairwiseAligner(mode="global", match=1, mismatch=0, open_gap_score=-1, extend_gap_score=-0.0)
[print(a.format()) for a in new.align(seq1,seq2)]
# Old behavior
# --DQ---------------------------
#   ||
# TEDQRNEEKAQREANKKIDQKQLQKDKQVYR
#   Score=0
#
# ------------------DQ-----------
#                   ||
# TEDQRNEEKAQREANKKIDQKQLQKDKQVYR
#   Score=0
#
# New behavior
# ------------------DQ-----------
# ------------------||-----------
# TEDQRNEEKAQREANKKIDQKQLQKDKQVYR
#
# --DQ---------------------------
# --||---------------------------
# TEDQRNEEKAQREANKKIDQKQLQKDKQVYR
```

### Expected behaviour

Same order of equally scored alignments between pairwise2 and Align

### Actual behaviour

Order is inverted

### Steps to reproduce

see the above code.

### Some comments
Obviously the above is a constructed test-case that happened to produce that inversion. Playing around (non-systematically) with sequences
```
seq1 = 'MGQPGNGSAF'
seq2 = seq1+seq1
```
yields the same order for both the old and the new methods, in this case aligning against the second half of seq2 and then against the first one.
```
Old behavior
----------MGQPGNGSAF
          ||||||||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

M----------GQPGNGSAF
|          |||||||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MG----------QPGNGSAF
||          ||||||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQ----------PGNGSAF
|||          |||||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQP----------GNGSAF
||||          ||||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPG----------NGSAF
|||||          |||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPGN----------GSAF
||||||          ||||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPGNG----------SAF
|||||||          |||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPGNGS----------AF
||||||||          ||
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPGNGSA----------F
|||||||||          |
MGQPGNGSAFMGQPGNGSAF
  Score=9

MGQPGNGSAF----------
||||||||||          
MGQPGNGSAFMGQPGNGSAF
  Score=9

New behavior
----------MGQPGNGSAF
----------||||||||||
MGQPGNGSAFMGQPGNGSAF

M----------GQPGNGSAF
|----------|||||||||
MGQPGNGSAFMGQPGNGSAF

MG----------QPGNGSAF
||----------||||||||
MGQPGNGSAFMGQPGNGSAF

MGQ----------PGNGSAF
|||----------|||||||
MGQPGNGSAFMGQPGNGSAF

MGQP----------GNGSAF
||||----------||||||
MGQPGNGSAFMGQPGNGSAF

MGQPG----------NGSAF
|||||----------|||||
MGQPGNGSAFMGQPGNGSAF

MGQPGN----------GSAF
||||||----------||||
MGQPGNGSAFMGQPGNGSAF

MGQPGNG----------SAF
|||||||----------|||
MGQPGNGSAFMGQPGNGSAF

MGQPGNGS----------AF
||||||||----------||
MGQPGNGSAFMGQPGNGSAF

MGQPGNGSA----------F
|||||||||----------|
MGQPGNGSAFMGQPGNGSAF

MGQPGNGSAF----------
||||||||||----------
MGQPGNGSAFMGQPGNGSAF
```
Also note, the choice of local vs global alignment here is not really well-thought out, again, I'm just curious about the test-case

Thanks a lot!

**Comment by mdehoon** [comment]
I am not sure what the question exactly is. In any case, `pairwise2` uses an iterative algorithm with a stack, whereas `PairwiseAligner` only knows the previously returned alignment and finds the next alignment from it. There is no guarantee that the alignments will be returned in the same order.

**Comment by MarkusPiotrowski** [comment]
To give a pragmatic advice: Since `pairwise2` is deprecated, you should find a way to deal with the different order of alignments.

To explain a little bit: The order of alignments has no meaning, in the code it is the question: If I have two or more possible traces of equal score, which one should I follow first? This is an arbitrary choice. From the point of the algorithm, all alignments have the same score, so none of them is 'better' than the other. Also, since `pairwise2` is restricted to 1000 (initial) alignments, it could happen that 'Align' returns an alignment that is not returned by `pairwise2`. 

What is your use-case that you want them to be in the same order? If it has some 'scientific' reason, i.e. you believe that the first alignments of `pairwise2` are for some reason 'better', you may think about changing the alignment parameters to favor such alignments. 

You may try to change the order of sequences and reverse them after the alignment was done. Not sure, if this gives the expected result.

**Comment by gph82** [comment]
Hi, thanks for both answers!

First off, this is almost a non-issue that I was curious about because my hard-coded test-suite flagged it up, as it was expecting the "old" order (in a bogus example).

Of course, 100% agree that the production code should deal with the new behavior. Also 100% agree that both alignments are equally fit (or any alignments with the same score). So, no, I don't have a particular reason to prefer the "first" one. 

For context, my use case is the following:  I have a known query sequence, and I'm trying to find pieces of it in an unknown target sequence to identify chains (or even segments) in a molecular dynamics (MD) topology. This is trivial when you have only one MD setup to keep track of, but I work on multiple projects with multiple MD setups. Each MD setup differs in the order of the molecular components and chain annotations. So it's hard keep track what chains/fragments corresponds to what protein in what setup. I'd rather align to reference sequences  and find it out automatically. In the case of GPCRs, it has the added bonus of mapping consensus residue labels. It looks like [this](https://proteinformatics.uni-leipzig.de/mdciao/notebooks/Tutorial.html) and then Ctrl+F  'align best'
```
[...]these fragments were found
fragment      0 with    354 AAs     LEU4 (     0) -   LEU394 (353   ) (0)  resSeq jumps
fragment      1 with    340 AAs     GLN1 (   354) -   ASN340 (693   ) (1)
fragment      2 with     66 AAs     ALA2 (   694) -    PHE67 (759   ) (2)
fragment      3 with    283 AAs    GLU30 (   760) -   LEU340 (1042  ) (3)  resSeq jumps
fragment      4 with      1 AAs   P0G395 (  1043) -   P0G395 (1043  ) (4)
No local file ./adrb2_human.xlsx found, checking online in
https://gpcrdb.org/services/residues/extended/adrb2_human ...done!
[...]
GPCR-labels align best with fragments: [3] (first-last: GLU30-LEU340).
These are the GPCR fragments mapped onto your topology:
    TM1 with     32 AAs    GLU30@1.29x29   (   760) -    PHE61@1.60x60   (791   ) (TM1)
   ICL1 with      4 AAs    GLU62@12.48x48  (   792) -    GLN65@12.51x51  (795   ) (ICL1)
    TM2 with     32 AAs    THR66@2.37x37   (   796) -    LYS97@2.68x67   (827   ) (TM2)
```

Other unit tests with much more realistic cases are passing no problem with `Align` (as they should be, I guess) but that toy example got flagged. 

Finally, and since you brought it up, I had also noticed the 1000 limit, since in elsewhere in my test suite I cycled over all alignments and it got stuck in the following test:
```
seq1='EVWVVGMGIVMSLIVLAIVFGNVLVITAIAKFERLQTVTNYFITSLACADLVMGLAVVPFGAAHILMKMWTFGNFWCEFWTSIDVLCVTASIETLCVIAVDRYFAITSPFKYQSLLTKNKARVIILMVWIVSGLTSFLPIQMHWYRATHQEAINCYANETCCDFFTNQAYAIASSIVSFYVPLVIMVFVYSRVFQECKRQLQKFCLKEHKALKTLGIIMGTFTLCWLPFFIVNIVHVIQDNLIRKEVYILLNWIGYVNSGFNPLIYCRSPDFRIAFQELLA'
seq2='AALGNSKTEDQRNEEKAQREANKKIEKQLQKDKQVYRATHRLLLLGAGESGKSTIVKQMRILHVNGFNGDGEKATKVQDIKNNLKEAIETIVAAMSNLVPPVELANPENQFRVDYILSVMNVPDFDFPPEFYEHAKALWEDEGVRACYERSNEYQLIDCAQYFLDKIDVIKQADYVPSDQDLLRCRVLTSGIFETKFQVDKVNFHMFDVGGQRDERRKWIQCFNDVTAIIFVVASSSYNVLIREDNQTNRLQEALNLFKSIWNNRWLRTISVILFLNKQDLLAEKVLAGKSKIEDYFPEFARYTTPEDATPEPGEDPRVTRAKYFIRDEFLRISTASGDGRHYCYPHFTCAVDTENIRRVFNDCRDIIQRMHLRQYELL'
old = pairwise2.align.globalms(seq1,seq2,1,0,-1,0)
new = Align.PairwiseAligner(match=1, mismatch=0, open_gap_score=-1, extend_gap_score=0).align(seq1,seq2)
print("n 'old' alignments:",len(old))
# n 'old' alignments: 1000 <-OK
print("n 'new' alignments:",len(new))
# n 'new' alignments: 5175997765056000 <- LOL
```
Since very little sequence identity but I'm _forcing_ it to get something, I get 10¹⁵ bad alignments. I'll enforce the 1000 limit myself, lol.

Thanks for your replies and feel free to close the issue. And thanks for maintaining BioPythoon, of course :)

**Comment by mdehoon** [comment]
> I get 10¹⁵ bad alignments.

Usually this happens when the alignment parameters are too loose. I would try and be more stringent with the gap scores, in particular the extend gap score, which is now set to 0, and maybe also set the mismatch score to -1. You can print a few of the 5175997765056000 alignments and have a look at them. Probably they are all very similar to each other, with just trivial differences between them.

**Comment by gph82** [comment]
In my use-case, having a high-number of equally scored alignments means poor alignment, since there's many many more "bad" alignments than "good" alignments.
 
Anyway I did play around with these parameters a bit, but it was not worth trying to get the right parameters that would cover all situations. There's a zoo of MD setups and PDB entries with unmodelled (missing) parts, inserted nanobodies, fused peptides and more that make e.g. fudging with the gap score nontrivial. 

In the end (when I implemented this originally), rather than parameter-explore to death until I "got it right in one pass" it was much more effective to do the alignment in two steps, where first I'm using the global alignment as a "screening tool" for target sub-fragments that highly likely match the query sequence and then I perform a second alignment only in those candidates. High sequence identity (=robust alignments no matter what parameters) _is_ expected once you limit yourself to the plausible candidates fragments.

Again, thanks for all the helpful comments!
